Add policy listener support

jrajahalme opened the pull request:
Add support for `listener` keyword in CNP to redirect matching traffic to an Envoy listener.

Please review commit-by-commit.

```release-note
Traffic can now we redirected to Envoy listeners via Cilium Network Policy `listener` option.
```

jrajahalme committed:
policy: Add ParserTypeCRD

Add a sticky parser type for CiliumEnvoyConfig CRDs. This will be used
for policy based redirect to custom Envoy listeners.

While CRD parser type will be redirected to Envoy, it is generally
handled by a custom Listener, which may not perform HTTP policy
enforcement. Thus this parser type is incompatible with HTTP rules.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Precompute redirect type on Attach()

Optimize policy processing by precomputing redirect types for the policy
when ready to avoid multiple scans of the policy filters later.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Cache redirect status on PerSelectorPolicy

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add CRD Listener support for egress policies

Add `listener` field to CNP and CCNP, that causes traffic to the
specified port(s) to be redirected to the named Envoy listener. If the
listener does not exist the traffic is not allowed at all.

When `listener.envoyConfig.kind` is left out it defaults to namespaced
`CiliumEnvoyConfig` for rules in namespaced policies (CNP) or to
cluster-scoped `CiliumClusterwideEnvoyConfig` for rules in cluster-scoped
policies (CCNP). Namespaced policies can also refer to cluster-scoped
listeners with an explicit `listener.envoyConfig.kind:
CiliumClusterwideEnvoyConfig`. Cluster-scoped policies can not refer to
namespaced listeners.

Endpoint policies are regenerated whenever Envoy listeners change to
update potential listener redirections in the bpf policy maps.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Inject Cilium Envoy filters to CEC with TcpProxy filter

Support injecting Cilium Network filter also to TcpProxy filter
chains. This enables SNI and other L3/L4 policy enforcement features also
on custom Envoy listeners applied via Cilium Envoy Config resources that
contain Envoy filter chains using TcpProxy filter, in addition to Http
Connection Manager filter.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Configure CEC Listners without L7LB for policy redirection

Tell envoy.ParseResources() if the resources are intended for a L7 load
balancer. If so, the Envoy filter chain is configured to not use original
source address and use the socket mark as the source endpoint ID. This was
the current behavior with all CEC CRDs.

When the CEC CRD does not contain a frontend service that is redirected
to the defined listener, then the new 'isL7LB' flag is passed as
'false'. In this case the Envoy filter chain is configured to allow use
of the original source address, and mark the upstream socket with the
numeric source security identity. This is also the way in which Cilium
uses Envoy Listener filter chains when enforcing policy for HTTP, for
example. In this mode L4 LB is assumed to be taken place, if applicable,
before traffic is redirected to the Envoy listener.

Prior to this change CEC CRDs were mostly only usable for L7 LB uses,
such as with Cilium Ingress. After this change CEC CRDs without a Service
redirect can be used in place of Cilium policy enforcement filter chains,
if desired.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Wait for Cluster before Listeners

Listener config can fail if we don't wait for clusters to be configured
first, so wait for clusters to be configured if both clusters and
listeners are configured at the same time.

While failure like this was once seen on local testing for the custom
listener support in a CNP, similar failure could also happen for Cilium
Ingress and Gateway API.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
proxy: use egress in test as CRD proxy ports only work for egress for now

Specify proxy port as egress (or not-ingress) in test cases as the
datapath currently supports ProxyTypeCRD only for L7 LB which is always
on the egress path.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
proxy: Do not create Envoy redirects for listeners defined in CEC CRDs

Add a new no-op CRDRedirect type to be used with Envoy listeners defined
in CEC CRDs. In this case the listeners already exist and new Envoy
Listener resources do not need to be created for them.

This is needed for the forthcoming policy feature where policy can refer
to a Listener defined in CEC CRD.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
proxy: Find CRD proxy ports by name instead of by type

Find CRD proxy port by name instead of type. This is needed for enabling
CEC CRD defined listeners to be used in CNPs. Prior to this CRD proxy
ports did not use this code path, which is only called from endpoint
policy updates, so there was no need to find CRD proxy ports by name.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

chancez commented on pkg/envoy/ciliumenvoyconfig.go:
```diff
@@ -63,9 +64,42 @@ type PortAllocator interface {
-func ParseResources(cecNamespace string, cecName string, anySlice []cilium_v2.XDSResource, validate bool, portAllocator PortAllocator) (Resources, error) {
+// ListenersAddedOrDeleted returns 'true' if a listener is added or removed when updating from 'old'
+// to 'new'
+func (old *Resources) ListenersAddedOrDeleted(new *Resources) bool {
```
Could probably use a unit test for this logic. Also, if we have a string-set data structure somewhere already, we could use it with listener names and do the set difference/intersection instead of manually calculating the difference with slices.

jrajahalme replied:
Will add a unit test, but generally I'd like to keep predicates like this not allocation memory when not needed. Not that this is on a particularly hot path, but still...

chancez commented on pkg/envoy/ciliumenvoyconfig_test.go:
```diff
@@ -551,6 +555,253 @@ func (s *JSONSuite) TestCiliumEnvoyConfigMulti(c *C) {
+                port_value: 443
+`
+
+func (s *JSONSuite) TestCiliumEnvoyConfigTCPProxyTermination(c *C) {
```
Is there a reason we can't just construct a full `Resource` and do a deep comparison instead of a dozen asserts on individual fields?

jrajahalme replied:
No reason, it was just easier.

jrajahalme committed:
envoy: Move resourceQualifiedName() to policy/api

Move resourceQualifiedName() to policy/api and export it so that it can
be used in policy as well as in envoy package.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

chancez commented on pkg/policy/api/rule_validation.go:
```diff
@@ -382,20 +382,32 @@ func (pr *PortRule) sanitize(ingress bool) error {
 
 	listener := pr.Listener
-	if listener != "" {
+	if listener != nil {
```
We could validate that if `listener.EnvoyConfig.Namespace` is empty for CNPs (should only be specified for CCNP).

chancez commented on pkg/policy/l4.go:
```diff
@@ -729,9 +729,13 @@ func createL4Filter(policyCtx PolicyContext, peerEndpoints api.EndpointSelectorS
 			l4.L7Parser = ParserTypeCRD
-			l4.Listener = pr.Listener
+			ns := pr.Listener.Namespace
+			if ns == "" {
```
Should probably check listener namespace matches policy namespace (for non-CCNP's)

chancez reviewed (changes requested):
I only reviewed recent commits so I may be missing something, but it seems we're not validating the Policies `listener.envoyConfig.namespace` matches the CNP namespace. 

Also, I'm a bit confused about the default of `"default"`. 

To clarify:
- Listener namespace should always be empty for CNPs referencing CECs, and the CNPs will reference CEC listeners in the same namespace.
- For CCNPs, the listener namespace should be required, and must match the namespace/name of an existing CEC.

I also just realized, we probably want to support CNPs referencing a CCEC. For that, we will need to add a `kind` field to the `listener.envoyConfig.kind` so we can disambiguate, similar to RoleBindings which can reference a Role or ClusterRole.

chancez reviewed (approved):
LGTM

christarazi commented on pkg/endpoint/bpf.go:
```diff
@@ -253,8 +253,13 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
+					// added or removed to reflect the change in listeners in
+					// the realized policy!
+					e.getLogger().WithError(err).Debug("Redirect rule skipped")
+					continue
```
Previously this was an error path, but now it's no longer returning errors here. Could you elaborate on the change of behavior here and what the consequences are? Is debug level enough for logging?

jrajahalme replied:
This is expected to happen when policy refers to a listener that does not exist yet, for example when restarting and k8s delivers the CNP before the related CEC. Related to this policy regeneration is now triggered when listeners are added/changed/deleted.

Debug level is enough to see that this happens as intended.

christarazi replied:
I think 

> for example when restarting and k8s delivers the CNP before the related CEC

is good to add to the code comment, because otherwise it's not clear why.

jrajahalme replied:
Reworded as:
```
	// Skip redirects that can not be created or updated.  This
	// can happen when a listener is missing, for example when
	// restarting and k8s delivers the CNP before the related
	// CEC.
	// Policy is regenerated when listeners are added or removed
	// to fix this condition when the listener is available.
	e.getLogger().WithField(logfields.Listener, l4.GetListener()).WithError(err).Debug("Redirect rule with missing listener skipped, will be applied once the listener is available")
	continue
```

christarazi commented on pkg/k8s/watchers/cilium_clusterwide_envoy_config.go:
```diff
@@ -113,6 +113,14 @@ func (k *K8sWatcher) addCiliumClusterwideEnvoyConfig(ccec *cilium_v2.CiliumClust
 	}
 
+	if len(resources.Listeners) > 0 {
+		// TODO: Policy does not need to be recomputed for this, but if we do not 'force'
+		// the bpf maps are not updated with the new proxy ports either. Move from the
+		// simple boolean to an enum that can more selectively skip regeneration steps (like
+		// we do for the datapath recompilations already?)
+		k.policyManager.TriggerPolicyUpdates(true, "Envoy Listeners added")
```
Do you intend to resolve this in this PR? Otherwise, we should create a GH issue for follow ups and put this item there, while referencing the GH issue in the code comment.

christarazi replied:
Also, I'm not following "Policy does not need to be recomputed for this, but if we do not 'force' the bpf maps are not updated with the new proxy ports either". AFAIU, policy needs to be recomputed in order for there to be a change to update in the BPF maps, so this `TriggerPolicyUpdates` sounds like it is needed in contrast to what the comment says :slightly_smiling_face:.

jrajahalme replied:
Not planning for this PR, as this related also to other cases where policy regeneration is forced. In this particular case we'd only need the existing endpoint policy be synced to bpf policy maps, as that is where the procy redirection port is resolved. This is why the endpoint policy itself would not need to be recomputed.

christarazi replied:
> In this particular case we'd only need the existing endpoint policy be synced to bpf policy maps, as that is where the procy redirection port is resolved.

So there are no BPF policymap updates needed here, but the proxy redirection port logic is embedded somewhere in the policy regeneration / recalculation path (therefore that's why a full policy recalculation works)? And we should only be triggering proxy redirection port logic and not a full regeneration / recalculation? How difficult is it to pull that out, just out of curiosity?

jrajahalme replied:
BPF policy maps need to be updated so that the previously skipped redirect rules are pushed to the bpf maps (see your first comment). We would need to push more context info with the regeneration trigger to be able to skip parts of the regeneration that are known to not be needed. We already do something similar to the BPF code generation, where we can skip compilation and only load the new programs, etc.

christarazi commented on pkg/policy/api/l4.go:
```diff
@@ -112,6 +112,39 @@ type TLSContext struct {
+
+	// Name is the name of the listener.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
```
If you don't expect this to be empty, then we can add validation on length so that it's rejected if empty.

[1]

jrajahalme replied:
Missed this, thought that would be part of required... ofcause it is not!

jrajahalme replied:
Added:
```
	// +kubebuilder:validation:MinLength=1
```

christarazi commented on pkg/policy/api/rule_validation.go:
```diff
@@ -381,6 +381,41 @@ func (pr *PortRule) sanitize(ingress bool) error {
+		}
+		if listener.EnvoyConfig.Name == "" {
+			return fmt.Errorf("Listener envoyConfig must have name (%s)", listener.Name)
+		}
```
[1] will allow you to drop this line.

jrajahalme replied:
Should I drop all other (manual) validation in favor of letting k8s worry about it as well?

jrajahalme replied:
Removed also other validation that is already done vie kubebuilder validation

christarazi commented on pkg/policy/api/rule_validation.go:
```diff
@@ -381,6 +381,41 @@ func (pr *PortRule) sanitize(ingress bool) error {
+		switch listener.EnvoyConfig.Kind {
+		case "CiliumEnvoyConfig", "CiliumClusterwideEnvoyConfig", "":
+		default:
+			return fmt.Errorf("Listener envoyConfig.kind must be CiliumEnvoyConfig, CiliumClusterwideEnvoyConfig, or left empty (%s)", listener.Name)
```
This error msg suggests that empty is not a valid `kind`, but the case above handles it. Also, the godoc on the `Kind` type above also suggests that it's optional (which means the field can be empty). So which is it? :slightly_smiling_face:

jrajahalme replied:
Empty is allowed, with sane defaulting as I think I wrote somewhere.

christarazi commented on pkg/policy/api/rule_validation.go:
```diff
@@ -381,6 +381,41 @@ func (pr *PortRule) sanitize(ingress bool) error {
+		// tested.
+		// TODO (jrajahalme): Lift this limitation for proxylib based parsers if needed and
+		// when tested.
+		if !pr.Rules.IsEmpty() {
```
Couple of TODOs here. Do you intend to implement in this PR? Apply follow up comment from above if not.

jrajahalme replied:
Added wording for "follow-up work" for both of these.

christarazi commented on pkg/policy/l4.go:
```diff
@@ -724,8 +727,34 @@ func createL4Filter(policyCtx PolicyContext, peerEndpoints api.EndpointSelectorS
+					// Cluster-scoped CCNP tries to use namespaced
+					// CiliumEnvoyConfig
+					//
+					// TODO: Catch this in rule validation once we have a
+					// validation context in there so that we can differentiate
+					// between CNP and CCNP at validation time.
```
Unfortunately, this may not be possible to do at CRD-level validation, because the CCNP type is embedded inside the CNP type. This means that the validation is simply copied over to CCNP. This has it's pros and cons, because 99% of cases, a CCNP should be validated like a CNP as they're pretty much the same thing, except when they are not. 

It's also not as easy as just not embedding the type because there are a lot of places in the code which assumes this structure, last I looked (>1y).

jrajahalme replied:
I meant `pkg/policy/api/rule_validation.go` with this, not k8s client-side validation. On our own policy API this would need passing in context w.r.t. what kind of policy in being validated. That is somewhat large refactor, so did not attempt it in this PR.

christarazi reviewed (commented):
Great work! A few comments below.
